### DISCUSSION TITLE
cell decorator not working as expected

### DISCUSSION THREAD

**Discussion by jonnyroth** [gdsfactory/gdsfactory#2628] [Q&A] [answered]
I'm having an issue where I believe my coded example is free of "dangerous intermediate cells" but multiple calls to a @gf.cell decorated function return the same value in subsequent calls, despite passing a different parameter.

The example below creates a cell, then slices it at different x values.   
With "ver=1" which is a cell with text, it works as expected:
![image](https://github.com/gdsfactory/gdsfactory/assets/24799153/c793cb3b-8468-49c2-9bd1-13e81889b95a)
With "ver=2" it uses a different cell, and the sliced geometry looks like whichever function call is made first:
![image](https://github.com/gdsfactory/gdsfactory/assets/24799153/4d6bcdcc-b9f5-4b21-bbce-a04f6643f0d4)
If I comment out the cell decorator on function "component_from_sliced_section" then it works fine:
![image](https://github.com/gdsfactory/gdsfactory/assets/24799153/868532c4-17ef-4249-9ff9-fbe5173393db)

Apologies for the example length.  If I removed more from my original layout code it didn't demonstrate the issue, and I'm not sure of the root cause.  I'm running v7.10.6. Thanks.

```python
import gdsfactory as gf
import gdstk

@gf.cell #If this cell decorator is removed, ver=2 works fine.
def component_from_sliced_section(sliced_section):
    print(sliced_section.__hash__()) #checking if different calls have identical hashes that could mess up cell decorator
    component=gf.Component()
    for polygon in sliced_section:
        component.add_polygon(polygon)
    return component

@gf.cell
def slice(in_cell: gf.Component, x_max=None):
    sliced_polygons=gdstk.slice(in_cell.flatten().polygons, position=x_max,axis='x')
    sliced_cell=component_from_sliced_section(tuple(sliced_polygons[0])) #i thought passing a tuple(hashable) would fix but does not.
    return sliced_cell

@gf.cell
def build_geometry(ver):
    if ver==1:
        return gf.components.text(text='THIS IS A TEST',size=300,layer=(1,0))
    if ver==2:
        component = gf.Component()
        path_top = gf.Path()
        path_top.append([gf.path.arc(radius=7000, angle=-16)])
        waveguide_top = gf.path.extrude(path_top, layer=(1,0),width=3)
        wg_instance_top = component << waveguide_top
        wg_instance_top.rotate(8)
        component.add_polygon([(-200,0),(200,0),(0,100)],layer=(1,0))
        return component

@gf.cell
def geo_with_slice(ver,x_max):
    cell1 = build_geometry(ver)
    cell2 = slice(cell1, x_max=x_max)
    return cell2

ver=2 #ver= 1 works as expected. ver= 2 does not.
top1=gf.Component(name='500_FIRST')
a= top1 << build_geometry(ver)
b= top1 << geo_with_slice(ver,500)
c= top1 << geo_with_slice(ver,1500)
a.movey(1000)
b.movey(500)
top1.show()

gf.clear_cache()

#reversing order of calls to geo_with_slice makes both calls return based on 1st call
top1=gf.Component(name='1500_FIRST')
a= top1 << build_geometry(ver)
c= top1 << geo_with_slice(ver,1500)
b= top1 << geo_with_slice(ver,500)
a.movey(1000)
b.movey(500)
top1.show()
```

**Comment by tvt173** [comment, marked as answer]
hi @jonnyroth 

yes, as you note in the comments of your example, there seems to be an issue with the serialization of the parameters for `component_from_sliced_section`. more specifically, it seems to be an issue with serialization of `gdstk.Polygon` objects as parameters. if you remove the `cell` decorator from that function, everything seems to work fine. 

i identified the issue and will submit a PR for it, but if you are in the middle of something and want to continue using the current version of gdsfactory you are pinned to, I would simply remove the decorator for that function. assuming you are only using it like you are using it within this example, it should be safe to do so, as you are re-returning and decorating it as a cell anyways within another function

**Reply by tvt173** [reply]
see PR https://github.com/gdsfactory/gdsfactory/pull/2629

**Reply by jonnyroth** [reply]
Thanks @tvt173 for the help & quick resolution!  Yes, for now removing the cell decorator fixed all the issues in my layout.
